File: tests/Hedgehog.Stateful.Tests.FSharp/SequentialCounterSpec.fs
module Hedgehog.Stateful.Tests.CounterFixtureClean

open Hedgehog.FSharp
open Hedgehog.Linq
open Hedgehog.Stateful
open Xunit
type Counter() =
    let mutable value = 0
    member _.Increment() = value <- value + 1
    member _.Decrement() = value <- value - 1
    member _.AddRandom() =
        let rnd = System.Random().Next(100)
        value <- value + rnd
        value
    member _.Set(n: int) =
        value <- n
    member _.Get() = value
    member _.Reset() = value <- 0

type CounterState = {
    CurrentCount: Var<int>  // Symbolic reference to current count
}

/// Increment command - SUT is a typed parameter!
type IncrementCommand() =
    inherit Command<Counter, CounterState, unit, int>()

    override _.Name = "Increment"
    override _.Precondition _ = true
    override _.Execute(counter, _, _, _) =
        counter.Increment()
        let result = counter.Get()  // Return new value
        System.Threading.Tasks.Task.FromResult(result)
    override _.Generate _ = Gen.constant ()
    override _.Update(_, _, outputVar) = { CurrentCount = outputVar }

    override _.Ensure(env, oldState, _, _, result) =
        let oldCount = oldState.CurrentCount.Resolve(env)
        result = oldCount + 1

type AddRandomCommand() =
    inherit Command<Counter, CounterState, unit, int>()

    override _.Name = "AddRandom"
    override _.Precondition _ = true
    override _.Execute(counter, _, _, _) = System.Threading.Tasks.Task.FromResult(counter.AddRandom())
    override _.Generate _ = Gen.constant ()
    override _.Update(_, _, outputVar) = { CurrentCount = outputVar }

/// Decrement command
type DecrementCommand() =
    inherit Command<Counter, CounterState, unit, int>()

    override _.Name = "Decrement"
    override _.Precondition _ = true
    override _.Execute(counter, _, _, _) =
        counter.Decrement()
        let result = counter.Get()  // Return new value
        System.Threading.Tasks.Task.FromResult(result)

    override _.Generate _ = Gen.constant ()
    override _.Update(_, _, outputVar) = { CurrentCount = outputVar }

    override _.Ensure(env, oldState, _, _, result) =
        let oldCount = oldState.CurrentCount.Resolve(env)
        result = oldCount - 1

/// Reset command
type ResetCommand() =
    inherit Command<Counter, CounterState, unit, int>()

    override _.Name = "Reset"
    override _.Precondition _ = true
    override _.Execute(counter, _, _, _) =
        counter.Reset()
        let result = counter.Get()  // Return new value (should be 0)
        System.Threading.Tasks.Task.FromResult(result)

    override _.Generate _ = Gen.constant ()
    override _.Update(_, _, outputVar) = { CurrentCount = outputVar }

    override _.Ensure(_, _, _, _, result) =
        // Reset always sets to 0
        result = 0

/// Set command - demonstrates the bug detection
type SetCommand() =
    inherit Command<Counter, CounterState, int, int>()

    override _.Name = "Set"
    override _.Precondition _ = true
    override _.Execute(counter, _, _, value) =
        counter.Set(value)
        let result = counter.Get()
        System.Threading.Tasks.Task.FromResult(result)
    override _.Generate _ = Gen.int32 (Range.linearFrom 0 -10 100)
    override _.Update(_, _, outputVar) = { CurrentCount = outputVar }

    override _.Ensure(_, _, _, input, result) =
        Assert.Equal(result, input)
        result = input

/// Get command - returns a value
type GetCommand() =
    inherit Command<Counter, CounterState, unit, int>()

    override _.Name = "Get"
    override _.Precondition _ = true
    override _.Execute(counter, _, _, _) = System.Threading.Tasks.Task.FromResult(counter.Get())
    override _.Generate _ = Gen.constant ()
    override _.Update(_, _, outputVar) = { CurrentCount = outputVar }

    override _.Ensure(env, oldState, _, _, result) =
        result = oldState.CurrentCount.Resolve(env)

/// SequentialSpecification that manages Counter lifecycle
type CounterSpec() =
    inherit SequentialSpecification<Counter, CounterState>()

    override _.SetupCommands = [|
        ResetCommand()
    |]

    // Initial model state - unbound var with default value 0
    override _.InitialState = { CurrentCount = Var.symbolic 0 }

    // Generate sequences of 1-50 actions
    override _.Range = Range.linear 1 50

    // Commands - no SUT needed in constructors!
    override _.Commands = [|
        IncrementCommand()
        DecrementCommand()
        ResetCommand()
        GetCommand()
        SetCommand()
        AddRandomCommand()
    |]


[<Fact>]
let ``Counter test with clean SUT parameter API``() =
    let sut = Counter()
    // Use CheckWith to create a fresh Counter for each property test run
    CounterSpec().ToProperty(sut).Check()

File: tests/Hedgehog.Stateful.Tests.CSharp/SequentialCounterSpec.cs
using Hedgehog.Stateful.Linq;

namespace Hedgehog.Stateful.Tests.CSharp;

using Hedgehog.Linq;
using Xunit;

/// <summary>
/// Simple counter for demonstrating sequential state machine testing
/// </summary>
public class Counter
{
    private int _value;

    public void Increment() => _value++;

    public void Decrement() => _value--;

    public int AddRandom()
    {
        var rnd = new Random().Next(100);
        _value += rnd;
        return _value;
    }

    public void Set(int n) => _value = n;

    public int Get() => _value;

    public void Reset() => _value = 0;
}

/// <summary>
/// Model state tracking the current count symbolically
/// </summary>
public record CounterState
{
    public required Var<int> CurrentCount { get; init; }
}

/// <summary>
/// Increment command - increments the counter and returns new value
/// </summary>
public class IncrementCommand : Command<Counter, CounterState, NoValue, int>
{
    public override string Name => "Increment";

    public override bool Precondition(CounterState state) => true;
    public override bool Require(Env env, CounterState state, NoValue value) => Precondition(state);

    public override Task<int> Execute(Counter sut, Env env, CounterState state, NoValue value)
    {
        sut.Increment();
        var result = sut.Get();
        return Task.FromResult(result);
    }

    public override Gen<NoValue> Generate(CounterState state) =>
        Gen.Constant(NoValue.Value);

    public override CounterState Update(CounterState state, NoValue value, Var<int> outputVar) =>
        state with { CurrentCount = outputVar };

    public override bool Ensure(Env env, CounterState oldState, CounterState newState, NoValue value, int result)
    {
        var oldCount = oldState.CurrentCount.Resolve(env);
        return result == oldCount + 1;
    }
}

/// <summary>
/// AddRandom command - adds a random value to the counter
/// </summary>
public class AddRandomCommand : Command<Counter, CounterState, NoValue, int>
{
    public override string Name => "AddRandom";

    public override bool Precondition(CounterState state) => true;
    public override bool Require(Env env, CounterState state, NoValue value) => Precondition(state);

    public override Task<int> Execute(Counter sut, Env env, CounterState state, NoValue value) =>
        Task.FromResult(sut.AddRandom());

    public override Gen<NoValue> Generate(CounterState state) =>
        Gen.Constant(NoValue.Value);


    public override CounterState Update(CounterState state, NoValue value, Var<int> outputVar) =>
        state with { CurrentCount = outputVar };

    public override bool Ensure(Env env, CounterState oldState, CounterState newState, NoValue value, int output) => true;
}

/// <summary>
/// Decrement command - decrements the counter and returns new value
/// </summary>
public class DecrementCommand : Command<Counter, CounterState, NoValue, int>
{
    public override string Name => "Decrement";

    public override bool Precondition(CounterState state) => true;
    public override bool Require(Env env, CounterState state, NoValue value) => Precondition(state);

    public override Task<int> Execute(Counter sut, Env env, CounterState state, NoValue value)
    {
        sut.Decrement();
        var result = sut.Get();
        return Task.FromResult(result);
    }

    public override Gen<NoValue> Generate(CounterState state) =>
        Gen.Constant(NoValue.Value);


    public override CounterState Update(CounterState state, NoValue value, Var<int> outputVar) =>
        state with { CurrentCount = outputVar };

    public override bool Ensure(Env env, CounterState oldState, CounterState newState, NoValue value, int result)
    {
        var oldCount = oldState.CurrentCount.Resolve(env);
        return result == oldCount - 1;
    }
}

/// <summary>
/// Reset command - resets the counter to 0
/// </summary>
public class ResetCommand : Command<Counter, CounterState, NoValue, int>
{
    public override string Name => "Reset";

    public override bool Precondition(CounterState state) => true;
    public override bool Require(Env env, CounterState state, NoValue value) => Precondition(state);

    public override Task<int> Execute(Counter sut, Env env, CounterState state, NoValue value)
    {
        sut.Reset();
        var result = sut.Get();
        return Task.FromResult(result);
    }

    public override Gen<NoValue> Generate(CounterState state) =>
        Gen.Constant(NoValue.Value);


    public override CounterState Update(CounterState state, NoValue value, Var<int> outputVar) =>
        state with { CurrentCount = outputVar };

    public override bool Ensure(Env env, CounterState oldState, CounterState newState, NoValue value, int result)
    {
        // Reset always sets to 0
        return result == 0;
    }
}

/// <summary>
/// Set command - sets the counter to a specific value
/// </summary>
public class SetCommand : Command<Counter, CounterState, int, int>
{
    public override string Name => "Set";

    public override bool Precondition(CounterState state) => true;
    public override bool Require(Env env, CounterState state, int input) => true;

    public override Task<int> Execute(Counter sut, Env env, CounterState state, int value)
    {
        sut.Set(value);
        var result = sut.Get();
        return Task.FromResult(result);
    }

    public override Gen<int> Generate(CounterState state) =>
        Gen.Int32(Range.LinearFromInt32(0, -10, 100));

    public override CounterState Update(CounterState state, int input, Var<int> outputVar) =>
        state with { CurrentCount = outputVar };

    public override bool Ensure(Env env, CounterState oldState, CounterState newState, int input, int result)
    {
        Assert.Equal(input, result);
        return result == input;
    }
}

/// <summary>
/// Get command - returns the current counter value
/// </summary>
public class GetCommand : Command<Counter, CounterState, NoValue, int>
{
    public override string Name => "Get";

    public override bool Precondition(CounterState state) => true;
    public override bool Require(Env env, CounterState state, NoValue value) => Precondition(state);

    public override Task<int> Execute(Counter sut, Env env, CounterState state, NoValue value) =>
        Task.FromResult(sut.Get());

    public override Gen<NoValue> Generate(CounterState state) =>
        Gen.Constant(NoValue.Value);

    public override CounterState Update(CounterState state, NoValue value, Var<int> outputVar) =>
        state with { CurrentCount = outputVar };

    public override bool Ensure(Env env, CounterState oldState, CounterState newState, NoValue value, int result)
    {
        return result == oldState.CurrentCount.Resolve(env);
    }
}

/// <summary>
/// Sequential specification that manages Counter lifecycle
/// </summary>
public class CounterSpec : SequentialSpecification<Counter, CounterState>
{
    public override ICommand<Counter, CounterState>[] SetupCommands => [new ResetCommand()];

    public override ICommand<Counter, CounterState>[] CleanupCommands => [
        new ResetCommand()
    ];

    public override CounterState InitialState => new() { CurrentCount = Var.Symbolic(0) };

    public override Range<int> Range => Hedgehog.Linq.Range.LinearInt32(1, 50);

    public override ICommand<Counter, CounterState>[] Commands =>
        [
            new IncrementCommand(),
            new DecrementCommand(),
            new ResetCommand(),
            new GetCommand(),
            new SetCommand(),
            new AddRandomCommand()
        ];
}

public class SequentialCounterTests
{
    [Fact]
    public void CounterTest()
    {
        var sut = new Counter();
        new CounterSpec().ToProperty(sut).Check();
    }
}
